cosim/emulator/src/vbridge_impl.cc

```diff
@@ -0,0 +1,476 @@
+#include <fmt/core.h>
+#include <glog/logging.h>
+
+#include "disasm.h"
+
+#include "verilated.h"
+
+#include "exceptions.h"
+#include "vbridge_impl.h"
+#include "util.h"
+#include "tl_interface.h"
+#include "simple_sim.h"
+#include "vpi.h"
+
+#include "glog_exception_safe.h"
+
+/// convert TL style size to size_by_bytes
+inline uint32_t decode_size(uint32_t encoded_size) {
+  return 1 << encoded_size;
+}
+
+VBridgeImpl::VBridgeImpl() :
+    sim(1 << 30),
+    isa("rv64gc", "msu"),
+    _cycles(100),
+    proc(
+        /*isa*/ &isa,
+        /*varch*/ fmt::format("").c_str(),
+        /*sim*/ &sim,
+        /*id*/ 0,
+        /*halt on reset*/ true,
+        /* endianness*/ memif_endianness_little,
+        /*log_file_t*/ nullptr,
+        /*sout*/ std::cerr){
+}
+
+void VBridgeImpl::setup(const std::string &_bin, const std::string &_ebin,const std::string &_wave, uint64_t _reset_vector, uint64_t cycles) {
+  this->bin = _bin;
+  this->ebin = _ebin;
+  this->wave = _wave;
+  this->reset_vector = _reset_vector;
+  this->timeout = cycles;
+}
+
+
+
+void VBridgeImpl::reset() {
+  top.clock = 0;
+  top.reset = 1;
+  top.resetVector = 0x1000;
+  top.eval();
+  tfp.dump(0);
+
+  // posedge
+  top.clock = 1;
+  top.eval();
+  tfp.dump(1);
+
+
+  // negedge
+  top.reset = 0;
+  top.clock = 0;
+  top.eval();
+  tfp.dump(2);
+  // posedge
+  LOG(INFO) << fmt::format("Reset compeleted, now we start");
+  top.reset = 0;
+  top.clock = 1;
+  top.eval();
+  tfp.dump(3);
+  ctx.time(2);
+}
+
+VBridgeImpl::~VBridgeImpl() {
+  terminate_simulator();
+}
+
+void VBridgeImpl::configure_simulator(int argc, char **argv) {
+  ctx.commandArgs(argc, argv);
+}
+
+void VBridgeImpl::init_spike() {
+  auto state = proc.get_state();
+  LOG(INFO) << fmt::format("Spike reset misa={:08X}",state->misa->read());
+  LOG(INFO) << fmt::format("Spike reset mstatus={:08X}",state->mstatus->read());
+  // load binary to reset_vector
+  sim.load(bin,ebin, reset_vector);
+}
+
+SpikeEvent *find_se_to_issue();void VBridgeImpl::init_simulator() {
+  Verilated::traceEverOn(true);
+  top.trace(&tfp, 99);
+  tfp.open(wave.c_str());
+  _cycles = timeout;
+}
+
+void VBridgeImpl::terminate_simulator() {
+  if (tfp.isOpen()) {
+    tfp.close();
+  }
+  top.final();
+}
+
+uint64_t VBridgeImpl::get_t() {
+  return ctx.time();
+}
+
+uint8_t VBridgeImpl::load(uint64_t address){
```

Comment:
The spacing style is not consistent with the original code. It includes the spacing between `)` and `{`, the spacing between `if`, `while`, `switch` and `(`

Reply:
Thanks for your advice. I have reformatted the codes with clang-format.